tonic-grpc-web/src/call/grpc_web_call.rs

```diff
@@ -0,0 +1,200 @@
+        let len = trailers.len();
+        assert!(len <= std::u32::MAX as usize);
+
+        let mut frame = Vec::with_capacity(len + HEADER_SIZE);
+        frame.push(GRPC_WEB_TRAILERS_MARKER);
+        frame.put_u32(len as u32);
+        frame.extend(trailers);
+
+        frame
+    }
+}
+
+impl<B> GrpcWebCall<B>
+where
+    B: Body<Data = Bytes> + Unpin,
+    B::Error: Error,
+{
+    fn poll_decode(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<B::Data, Status>>> {
+        match self.encoding {
+            Encoding::Base64 => loop {
+                // TODO: Do not loop until buf.remaining() % 4 == 0, emit a chunk as soon as possible.
+                if let Some(bytes) = self.decode_chunk()? {
+                    return Poll::Ready(Some(Ok(bytes)));
+                }
+
+                match ready!(Pin::new(&mut self.inner).poll_data(cx)) {
+                    Some(Ok(data)) => self.buf.put(data),
+                    Some(Err(e)) => return Poll::Ready(Some(Err(internal_error(e)))),
+                    None => {
+                        return if self.buf.has_remaining() {
+                            Poll::Ready(Some(Err(internal_error("malformed base64 request"))))
+                        } else {
+                            Poll::Ready(None)
+                        }
+                    }
+                }
+            },
+
+            Encoding::None => match ready!(Pin::new(&mut self.inner).poll_data(cx)) {
+                Some(res) => Poll::Ready(Some(res.map_err(internal_error))),
+                None => Poll::Ready(None),
+            },
+        }
+    }
+
+    fn poll_encode(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<B::Data, Status>>> {
+        if let Some(mut res) = ready!(Pin::new(&mut self.inner).poll_data(cx)) {
+            if self.encoding == Encoding::Base64 {
+                res = res.map(|b| base64::encode(b).into())
+            }
+
+            return Poll::Ready(Some(res.map_err(internal_error)));
+        }
+
+        // this flag is needed because the inner stream never
+        // returns Poll::Ready(None) when polled for trailers
+        if self.poll_trailers {
+            return match ready!(Pin::new(&mut self.inner).poll_trailers(cx)) {
+                Ok(Some(map)) => {
+                    let mut frame = self.make_trailers_frame(map);
+
+                    if self.encoding == Encoding::Base64 {
+                        frame = base64::encode(frame).into_bytes();
+                    }
+
+                    self.poll_trailers = false;
+                    Poll::Ready(Some(Ok(frame.into())))
+                }
+                Ok(None) => Poll::Ready(None),
+                Err(e) => Poll::Ready(Some(Err(internal_error(e)))),
+            };
+        }
+
+        Poll::Ready(None)
+    }
+}
+
+impl<B> Body for GrpcWebCall<B>
+where
+    B: Body<Data = Bytes> + Unpin,
```

Comment:
I can't remember but I don't think we need this `Unpin` requirement? We can project since we are pinned as well.